temporal-sdk/src/test/java/io/temporal/workflow/contextPropagatonTests/CustomActivityContextPropagationTest.java

```diff
@@ -0,0 +1,157 @@
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TestWatcher;
+import org.junit.runner.Description;
+import org.slf4j.MDC;
+
+public class CustomActivityContextPropagationTest {
+  public static final String CUSTOM_KEY = "custom_key";
+
+  @Rule
+  public TestWatcher watchman =
+      new TestWatcher() {
+        @Override
+        protected void failed(Throwable e, Description description) {
+          System.err.println(testEnvironment.getDiagnostics());
+        }
+      };
+
+  private static final String TASK_QUEUE = "test-workflow";
+
+  private TestWorkflowEnvironment testEnvironment;
+
+  @Before
+  public void setUp() {
+    TestEnvironmentOptions options =
+        TestEnvironmentOptions.newBuilder()
+            .setWorkflowClientOptions(
+                WorkflowClientOptions.newBuilder()
+                    .setContextPropagators(
+                        Arrays.asList(
+                            new TestContextPropagator(),
+                            // This has to be on worker config. Propagators applied on
+                            // ActivityOptions are not propagated automatically to the Activity Worker side.
+                            new TestContextPropagator(CUSTOM_KEY)))
+                    .build())
+            .build();
+    testEnvironment = TestWorkflowEnvironment.newInstance(options);
+  }
+
+  @After
+  public void tearDown() {
+    testEnvironment.close();
+  }
+
+  @WorkflowInterface
+  public interface TestWorkflow {
+    @WorkflowMethod
+    String workflow1(String input);
+  }
+
+  @ActivityInterface
+  public interface TestActivity {
+    String activity(String input);
+  }
+
+  public static class ContextActivityImpl implements TestActivity {
+    @Override
+    public String activity(String input) {
+      return "activity_" + MDC.get(TestContextPropagator.DEFAULT_KEY) + "_" + MDC.get(CUSTOM_KEY);
+    }
+  }
+
+  public static class ContextPropagationActivityWorkflowImpl implements TestWorkflow {
+    @Override
+    public String workflow1(String input) {
+      ActivityOptions options =
+          ActivityOptions.newBuilder()
+              .setScheduleToCloseTimeout(Duration.ofSeconds(5))
+              .setContextPropagators(
+                  Collections.singletonList(new TestContextPropagator(CUSTOM_KEY)))
+              .build();
+      TestActivity activity = Workflow.newActivityStub(TestActivity.class, options);
+      MDC.put(CUSTOM_KEY, "testing1234");
+
+      return activity.activity("foo");
+    }
+  }
+
+  public static class ContextPropagationLocalActivityWorkflowImpl implements TestWorkflow {
+    @Override
+    public String workflow1(String input) {
+      LocalActivityOptions options =
+          LocalActivityOptions.newBuilder()
+              .setScheduleToCloseTimeout(Duration.ofSeconds(5))
+              .setContextPropagators(
+                  Collections.singletonList(new TestContextPropagator(CUSTOM_KEY)))
```

Comment:
Specifying any custom context propagators this way doesn't make sense right now.
These propagators HAVE TO BE in WorkflowClientOptions and without that they will not be used for deserialization on the Activity Worker side.
If they are already in WorkflowClientOptions - they are used for serialization anyway for all activities, so specifying them on ActivityOptions doesn't change anything.